File: itinerary_compute/itinerary_builder.py
import numpy as np
import csv
from typing import List, Dict, Any, Optional
from content_based_interest import extract_interest_vector, cosine_similarity, recommend_activities


def load_distance_matrix(csv_path: str) -> Dict[str, Dict[str, float]]:
    """Load distance matrix from CSV file."""
    distances = {}
    with open(csv_path, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        for row in reader:
            activity_name = row['name']
            distances[activity_name] = {k: float(v) for k, v in row.items() if k != 'name'}
    return distances


def time_to_minutes(time_str: str) -> float:
    """Convert HH:MM to minutes since midnight."""
    hours, minutes = map(int, time_str.split(':'))
    return hours * 60 + minutes


def minutes_to_time(minutes: float) -> str:
    """Convert minutes since midnight to HH:MM."""
    hours = int(minutes // 60)
    mins = int(minutes % 60)
    return f"{hours:02d}:{mins:02d}"


def is_activity_feasible(
    activity: Dict[str, Any],
    current_time: float,
    end_time: float,
    current_location: str,
    distances: Dict[str, Dict[str, float]],
    return_point: str = "La Statue de Louis XIV"
) -> bool:
    """Check if activity is feasible given time constraints."""
    travel_time = distances.get(current_location, {}).get(activity['name'], 0)
    arrival_time = current_time + travel_time
    
    opening_time = activity['openingTime'] * 60
    closing_time = activity['closingTime'] * 60
    duration = activity['duration'] * 60
    
    # Check opening hours
    actual_start = max(arrival_time, opening_time)
    if actual_start >= closing_time:
        return False
    
    # Check if we can finish before closing
    finish_time = actual_start + duration
    if finish_time > closing_time:
        return False
    
    # Check if we can return in time
    return_time = distances.get(activity['name'], {}).get(return_point, 0)
    if finish_time + return_time > end_time:
        return False
    
    return True


def calculate_composite_score(
    activity: Dict[str, Any],
    current_time: float,
    current_location: str,
    cumulative_vector: np.ndarray,
    distances: Dict[str, Dict[str, float]],
    rec_scores: Dict[str, float],
    alpha: float = 1.0,
    beta: float = 0.4,
    gamma: float = 0.5,
    delta: float = 0.2
) -> float:
    """Calculate composite score for activity selection."""
    # Base recommendation score
    rec_score = rec_scores.get(str(activity['activityId']), 0)
    
    # Normalize travel time penalty
    travel_time = distances.get(current_location, {}).get(activity['name'], 0)
    max_travel = 35  # max reasonable travel time
    travel_penalty = min(travel_time / max_travel, 1.0)
    
    # Similarity penalty to visited activities
    activity_vector = extract_interest_vector(activity)
    similarity_penalty = 0.0
    if np.linalg.norm(cumulative_vector) > 0:
        similarity_penalty = max(0, cosine_similarity(activity_vector, cumulative_vector))
    
    # Timing bonus
    arrival_time = current_time + travel_time
    opening_time = activity['openingTime'] * 60
    wait_time = max(0, opening_time - arrival_time)
    timing_bonus = 1.0 - min(wait_time / 60, 1.0)  # normalize to 60 min max wait
    
    score = (
        alpha * rec_score
        - beta * travel_penalty
        - gamma * similarity_penalty
        + delta * timing_bonus
    )
    
    return score


def build_itinerary(
    likes_dislikes: List[Dict[str, Any]],
    activities: List[Dict[str, Any]],
    distances: Dict[str, Dict[str, float]],
    start_time: str,
    end_time: str,
    start_point: str = "La Statue de Louis XIV",
    max_activities: int = 10,
    alpha: float = 1.0,
    beta: float = 0.4,
    gamma: float = 0.5,
    delta: float = 0.2
) -> Dict[str, Any]:
    """
    Build optimized itinerary using greedy algorithm with composite scoring.
    
    Args:
        likes_dislikes: User swipe data
        activities: List of all activities
        distances: Distance matrix (minutes)
        start_time: Departure time (HH:MM)
        end_time: Return time (HH:MM)
        start_point: Starting location
        max_activities: Maximum number of activities
        alpha, beta, gamma, delta: Score weights
    
    Returns:
        Complete itinerary with timing and statistics
    """
    # Get recommendation scores
    scored_activities = recommend_activities(likes_dislikes, activities, exclude_seen=True)
    rec_scores = {str(a['activityId']): a['recommendation_score'] for a in scored_activities}
    
    # Initialize state
    current_time = time_to_minutes(start_time)
    end_time_min = time_to_minutes(end_time)
    current_location = start_point
    visited = set()
    cumulative_vector = np.zeros(9)
    itinerary = []
    
    # Add starting point
    itinerary.append({
        "order": 0,
        "activity_name": start_point,
        "arrival_time": minutes_to_time(current_time),
        "departure_time": minutes_to_time(current_time),
        "duration": 0,
        "waiting_time": 0,
        "travel_time_from_previous": 0
    })
    
    # Greedy selection loop
    while len(visited) < max_activities:
        # Filter candidates
        candidates = [
            act for act in scored_activities
            if str(act['activityId']) not in visited
            and act['name'] != start_point
            and is_activity_feasible(act, current_time, end_time_min, current_location, distances, start_point)
        ]
        
        if not candidates:
            break
        
        # Score all candidates
        scored_candidates = []
        for candidate in candidates:
            score = calculate_composite_score(
                candidate, current_time, current_location, cumulative_vector,
                distances, rec_scores, alpha, beta, gamma, delta
            )
            scored_candidates.append((candidate, score))
        
        # Select best
        best_activity, best_score = max(scored_candidates, key=lambda x: x[1])
        
        # Update state
        travel_time = distances.get(current_location, {}).get(best_activity['name'], 0)
        arrival_time = current_time + travel_time
        opening_time = best_activity['openingTime'] * 60
        actual_start = max(arrival_time, opening_time)
        waiting_time = max(0, opening_time - arrival_time)
        duration = best_activity['duration'] * 60
        departure_time = actual_start + duration
        
        # Add to itinerary
        itinerary.append({
            "order": len(itinerary),
            "activity_id": best_activity['activityId'],
            "activity_name": best_activity['name'],
            "arrival_time": minutes_to_time(arrival_time),
            "departure_time": minutes_to_time(departure_time),
            "duration": duration,
            "waiting_time": waiting_time,
            "travel_time_from_previous": travel_time,
            "composite_score": best_score,
            "recommendation_score": rec_scores.get(str(best_activity['activityId']), 0)
        })
        
        # Update state
        visited.add(str(best_activity['activityId']))
        cumulative_vector += extract_interest_vector(best_activity)
        current_location = best_activity['name']
        current_time = departure_time
    
    # Add return
    return_travel = distances.get(current_location, {}).get(start_point, 0)
    final_arrival = current_time + return_travel
    
    itinerary.append({
        "order": len(itinerary),
        "activity_name": start_point,
        "arrival_time": minutes_to_time(final_arrival),
        "departure_time": minutes_to_time(final_arrival),
        "duration": 0,
        "waiting_time": 0,
        "travel_time_from_previous": return_travel
    })
    
    # Calculate stats
    total_travel = sum(step['travel_time_from_previous'] for step in itinerary)
    total_visit = sum(step['duration'] for step in itinerary)
    total_waiting = sum(step['waiting_time'] for step in itinerary)
    
    return {
        "departure_time": start_time,
        "arrival_time": minutes_to_time(final_arrival),
        "total_duration": final_arrival - time_to_minutes(start_time),
        "total_activities": len(visited),
        "itinerary": itinerary,
        "stats": {
            "total_travel_time": total_travel,
            "total_visit_time": total_visit,
            "total_waiting_time": total_waiting
        }
    }


if __name__ == "__main__":
    import json
    
    # Load data
    distances = load_distance_matrix('../activity_distances_temp.csv')
    
    with open('../src/assets/data/activity_v2.json', 'r', encoding='utf-8') as f:
        activities = json.load(f)

    with open('../jb_visit.json', 'r', encoding='utf-8') as f:
        likes_dislikes = json.load(f)
    

    
    # Build itinerary
    result = build_itinerary(
        likes_dislikes=likes_dislikes,
        activities=activities,
        distances=distances,
        max_activities=100,
        start_time="09:00",
        end_time="18:00",
        alpha=1.0,
        beta=0.1,
        gamma=0.5,
        delta=0.2
    )
    
    # Display results
    print(f"=== Itinéraire optimisé ===")
    print(f"Départ: {result['departure_time']}")
    print(f"Retour: {result['arrival_time']}")
    print(f"Durée totale: {result['total_duration']:.0f} min")
    print(f"Activités visitées: {result['total_activities']}\n")
    
    for step in result['itinerary']:
        if step['order'] == 0 or step['order'] == len(result['itinerary']) - 1:
            print(f"{step['order']}. {step['activity_name']} - {step['arrival_time']}")
        else:
            print(f"{step['order']}. {step['activity_name']}")
            print(f"   Arrivée: {step['arrival_time']} | Départ: {step['departure_time']}")
            print(f"   Durée visite: {step['duration']:.0f}min | Trajet: {step['travel_time_from_previous']:.0f}min")
            print(f"   Score: {step['composite_score']:.3f}")
    
    print(f"\n=== Statistiques ===")
    print(f"Temps de visite: {result['stats']['total_visit_time']:.0f} min")
    print(f"Temps de trajet: {result['stats']['total_travel_time']:.0f} min")
    print(f"Temps d'attente: {result['stats']['total_waiting_time']:.0f} min")

File: src/lib/itinerary/itinerary-builder.ts
import {
  type Activity,
  type LikeDislike,
  extractInterestVector,
  cosineSimilarity,
  recommendActivities,
} from "./content-based-interest";

export interface DistanceMatrix {
  [activityName: string]: {
    [targetName: string]: number;
  };
}

export interface ItineraryStep {
  order: number;
  activity_id?: string;
  activity_name: string;
  arrival_time: string;
  departure_time: string;
  duration: number;
  waiting_time: number;
  travel_time_from_previous: number;
  composite_score?: number;
  recommendation_score?: number;
}

export interface Itinerary {
  departure_time: string;
  arrival_time: string;
  total_duration: number;
  total_activities: number;
  itinerary: ItineraryStep[];
  stats: {
    total_travel_time: number;
    total_visit_time: number;
    total_waiting_time: number;
  };
}

function timeToMinutes(timeStr: string): number {
  const [hours, minutes] = timeStr.split(":").map(Number);
  return (hours ?? 0) * 60 + (minutes ?? 0);
}

function minutesToTime(minutes: number): string {
  const hours = Math.floor(minutes / 60);
  const mins = Math.floor(minutes % 60);
  return `${hours.toString().padStart(2, "0")}:${mins.toString().padStart(2, "0")}`;
}

function applyWalkSpeed(distances: DistanceMatrix, walkSpeed: number): DistanceMatrix {
  const coefficient = Math.min(0.2, walkSpeed / 100);
  const speedMultiplier = 1 + coefficient;
  
  const adjusted: DistanceMatrix = {};
  for (const [from, targets] of Object.entries(distances)) {
    adjusted[from] = {};
    for (const [to, time] of Object.entries(targets)) {
      adjusted[from]![to] = time / speedMultiplier;
    }
  }
  return adjusted;
}

function isActivityFeasible(
  activity: Activity,
  currentTime: number,
  endTime: number,
  currentLocation: string,
  distances: DistanceMatrix,
  returnPoint = "La Statue de Louis XIV"
): boolean {
  const travelTime = distances[currentLocation]?.[activity.name] ?? 0;
  const arrivalTime = currentTime + travelTime;

  const openingTime = activity.openingTime * 60;
  const closingTime = activity.closingTime * 60;
  const duration = activity.duration * 60;

  const actualStart = Math.max(arrivalTime, openingTime);
  if (actualStart >= closingTime) return false;

  const finishTime = actualStart + duration;
  if (finishTime > closingTime) return false;

  const returnTime = distances[activity.name]?.[returnPoint] ?? 0;
  if (finishTime + returnTime > endTime) return false;

  return true;
}

function calculateCompositeScore(
  activity: Activity & { recommendation_score: number },
  currentTime: number,
  currentLocation: string,
  cumulativeVector: number[],
  distances: DistanceMatrix,
  timeInCurrentZone: number,
  alpha = 1.0,
  beta = 0.4,
  gamma = 0.5,
  delta = 0.2,
  epsilon = 1.5
): number {
  const recScore = activity.recommendation_score;

  // Distance penalty with exponential scaling
  const travelTime = distances[currentLocation]?.[activity.name] ?? 0;
  const normalizedDistance = travelTime / 30; // normalize by 30min reference
  const exponentialPenalty = Math.pow(normalizedDistance, 2); // quadratic penalty
  
  // Zone change bonus: every 120-180 minutes, encourage zone changes
  const zoneChangeThreshold = 120; // 2 hours
  const zoneChangeWindow = 180; // 3 hours
  let distanceBonusFactor = 0;
  
  if (timeInCurrentZone >= zoneChangeThreshold && timeInCurrentZone <= zoneChangeWindow) {
    // Encourage far activities to change zones
    if (travelTime > 15) {
      distanceBonusFactor = 0.5 * ((timeInCurrentZone - zoneChangeThreshold) / (zoneChangeWindow - zoneChangeThreshold));
    }
  } else if (timeInCurrentZone > zoneChangeWindow) {
    // Strong encouragement to change zones
    if (travelTime > 15) {
      distanceBonusFactor = 0.8;
    }
  }
  
  const travelPenalty = exponentialPenalty - distanceBonusFactor;

  // Similarity penalty to visited activities
  const activityVector = extractInterestVector(activity);
  let similarityPenalty = 0;
  const cumulativeNorm = Math.sqrt(cumulativeVector.reduce((s, v) => s + v * v, 0));
  if (cumulativeNorm > 0) {
    similarityPenalty = Math.max(0, cosineSimilarity(activityVector, cumulativeVector));
  }

  // Timing bonus
  const arrivalTime = currentTime + travelTime;
  const openingTime = activity.openingTime * 60;
  const waitTime = Math.max(0, openingTime - arrivalTime);
  const timingBonus = 1.0 - Math.min(waitTime / 60, 1.0);

  return alpha * recScore - beta * epsilon * travelPenalty - gamma * similarityPenalty + delta * timingBonus;
}

export function buildItinerary(
  likesDislikes: LikeDislike[],
  activities: Activity[],
  distances: DistanceMatrix,
  startTime: string,
  endTime: string,
  walkSpeed = 50,
  startPoint = "La Statue de Louis XIV",
  maxActivities = 10,
  alpha = 1.0,
  beta = 0.4,
  gamma = 0.5,
  delta = 0.2,
  epsilon = 1.5
): Itinerary {
  const adjustedDistances = applyWalkSpeed(distances, walkSpeed);

  const scoredActivities = recommendActivities(likesDislikes, activities, 1.0, 0.5, true);
  const recScores = new Map(scoredActivities.map(a => [a.activityId, a.recommendation_score]));

  let currentTime = timeToMinutes(startTime);
  const endTimeMin = timeToMinutes(endTime);
  let currentLocation = startPoint;
  const visited = new Set<string>();
  const cumulativeVector = new Array(10).fill(0);
  const itinerary: ItineraryStep[] = [];
  let timeInCurrentZone = 0; // Track time spent in current zone
  let lastZoneChangeTime = currentTime;

  itinerary.push({
    order: 0,
    activity_name: startPoint,
    arrival_time: minutesToTime(currentTime),
    departure_time: minutesToTime(currentTime),
    duration: 0,
    waiting_time: 0,
    travel_time_from_previous: 0,
  });

  while (visited.size < maxActivities) {
    const candidates = scoredActivities.filter(
      act =>
        !visited.has(act.activityId) &&
        act.name !== startPoint &&
        isActivityFeasible(act, currentTime, endTimeMin, currentLocation, adjustedDistances, startPoint)
    );

    if (candidates.length === 0) break;

    const scoredCandidates = candidates.map(candidate => {
      const score = calculateCompositeScore(
        candidate,
        currentTime,
        currentLocation,
        cumulativeVector,
        adjustedDistances,
        timeInCurrentZone,
        alpha,
        beta,
        gamma,
        delta,
        epsilon
      );
      return { activity: candidate, score };
    });

    const best = scoredCandidates.reduce((max, curr) => (curr.score > max.score ? curr : max));
    const bestActivity = best.activity;
    const bestScore = best.score;

    const travelTime = adjustedDistances[currentLocation]?.[bestActivity.name] ?? 0;
    const arrivalTime = currentTime + travelTime;
    const openingTime = bestActivity.openingTime * 60;
    const actualStart = Math.max(arrivalTime, openingTime);
    const waitingTime = Math.max(0, openingTime - arrivalTime);
    const duration = bestActivity.duration * 60;
    const departureTime = actualStart + duration;
    
    // Track zone changes: if travel time > 15min, consider it a zone change
    if (travelTime > 15) {
      timeInCurrentZone = 0;
      lastZoneChangeTime = currentTime;
    } else {
      timeInCurrentZone = currentTime - lastZoneChangeTime;
    }

    itinerary.push({
      order: itinerary.length,
      activity_id: bestActivity.activityId,
      activity_name: bestActivity.name,
      arrival_time: minutesToTime(arrivalTime),
      departure_time: minutesToTime(departureTime),
      duration,
      waiting_time: waitingTime,
      travel_time_from_previous: travelTime,
      composite_score: bestScore,
      recommendation_score: recScores.get(bestActivity.activityId) ?? 0,
    });

    visited.add(bestActivity.activityId);
    const activityVector = extractInterestVector(bestActivity);
    for (let i = 0; i < 10; i++) {
      cumulativeVector[i] += activityVector[i] ?? 0;
    }
    currentLocation = bestActivity.name;
    currentTime = departureTime;
  }

  const returnTravel = adjustedDistances[currentLocation]?.[startPoint] ?? 0;
  const finalArrival = currentTime + returnTravel;

  itinerary.push({
    order: itinerary.length,
    activity_name: startPoint,
    arrival_time: minutesToTime(finalArrival),
    departure_time: minutesToTime(finalArrival),
    duration: 0,
    waiting_time: 0,
    travel_time_from_previous: returnTravel,
  });

  const totalTravel = itinerary.reduce((sum, step) => sum + step.travel_time_from_previous, 0);
  const totalVisit = itinerary.reduce((sum, step) => sum + step.duration, 0);
  const totalWaiting = itinerary.reduce((sum, step) => sum + step.waiting_time, 0);

  return {
    departure_time: startTime,
    arrival_time: minutesToTime(finalArrival),
    total_duration: finalArrival - timeToMinutes(startTime),
    total_activities: visited.size,
    itinerary,
    stats: {
      total_travel_time: totalTravel,
      total_visit_time: totalVisit,
      total_waiting_time: totalWaiting,
    },
  };
}
